core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1300,18 +1303,27 @@ class UnifiedLog(@volatile var logStartOffset: Long,
       } else if (targetTimestamp == ListOffsetsRequest.EARLIEST_LOCAL_TIMESTAMP) {
         val curLocalLogStartOffset = localLogStartOffset()
 
-        val earliestLocalLogEpochEntry = leaderEpochCache.asJava.flatMap(cache => {
+        val epochOpt: Optional[Integer] = leaderEpochCache.asJava.flatMap(cache => {
           val epoch = cache.epochForOffset(curLocalLogStartOffset)
-          if (epoch.isPresent) cache.epochEntry(epoch.getAsInt) else Optional.empty[EpochEntry]()
+          if (epoch.isPresent) Optional.of[Integer](epoch.getAsInt) else Optional.empty[Integer]()
         })
 
-        val epochOpt = if (earliestLocalLogEpochEntry.isPresent && earliestLocalLogEpochEntry.get().startOffset <= curLocalLogStartOffset)
-          Optional.of[Integer](earliestLocalLogEpochEntry.get().epoch)
-        else Optional.empty[Integer]()
-
```

Comment:
I didn't really see a point in this check `earliestLocalLogEpochEntry.get().startOffset <= curLocalLogStartOffset`. As far as I can tell the `cache.epochForOffset` already carries it out. Let me know in case I have misunderstood something.

Reply:
subjective:

Can we update the method similar to below for readability/clarity:

```java
if (remoteLogEnabled()) {
  val curHighestRemoteOffset = highestOffsetInRemoteStorage()
  var result: Optional[Integer] = Optional.of(RecordBatch.NO_PARTITION_LEADER_EPOCH)
  if (leaderEpochCache.isDefined) {
    val epochOpt = leaderEpochCache.get.epochForOffset(curHighestRemoteOffset)
    if (epochOpt.isPresent) {
      result = Optional.of(epochOpt.getAsInt) 
    }
  }
  Some(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, highestOffsetInRemoteStorage(), result))
}
```

Note when the `highestOffsetInRemoteStorage` is -1, then there won't be corresponding leaderEpoch, we have to return the `NO_PARTITION_LEADER_EPOCH` constant which is -1.

Reply:
Hopefully I have achieved both in the subsequent commit (reverted one of the changes and made both easier to read). Let me know if this isn't the case!